File: services/scraper/app/extractors/stores/registry.py
import json
import re
from urllib.parse import urlparse

from lxml import etree, html

from app.extractors.normalization import (
    clean_text,
    normalize_currency,
    normalize_discount,
    normalize_image_url,
    normalize_price,
)
from app.extractors.result import ExtractionResult
from app.extractors.stores.complex import (
    extract_aliexpress,
    extract_amazon,
    extract_asos,
    extract_flipkart,
    extract_foxtrot,
    extract_n11,
    extract_target,
    extract_trendyol,
    extract_zalando,
)
from app.extractors.stores.enrichment import (
    extract_rozetka_html,
    extract_shopgoodwill_html,
)
from app.extractors.stores.embedded import (
    extract_embedded_store,
    supports_embedded_store,
)
from app.extractors.stores.profiles import PROFILES, FieldRule, StoreProfile
from app.models import ProductData


def supported_store_patterns() -> tuple[str, ...]:
    return tuple(pattern for profile in PROFILES for pattern in profile.patterns) + (
        "bandcamp.com",
        "amazon.",
        "aliexpress.",
        "foxtrot.com.ua",
        "n11.com",
        "rozetka.com.ua",
        "shopgoodwill.com",
        "wildberries.ru",
    )


def get_store_profile(page_url: str) -> StoreProfile | None:
    hostname = (urlparse(page_url).hostname or "").lower()
    return next(
        (
            candidate
            for candidate in PROFILES
            if any(pattern in hostname for pattern in candidate.patterns)
        ),
        None,
    )


def extract_store_product(
    html_text: str,
    page_url: str,
) -> ExtractionResult | None:
    hostname = (urlparse(page_url).hostname or "").lower()
    if supports_embedded_store(page_url):
        embedded = extract_embedded_store(html_text, page_url)
        if embedded.product.title or embedded.product.price or embedded.product.image:
            return embedded
    complex_extractors = (
        ("amazon.", extract_amazon),
        ("aliexpress.", extract_aliexpress),
        ("asos.com", extract_asos),
        ("foxtrot.com.ua", extract_foxtrot),
        ("n11.com", extract_n11),
        ("target.com", extract_target),
        ("flipkart.com", extract_flipkart),
        ("trendyol.com", extract_trendyol),
        ("zalando.", extract_zalando),
        ("rozetka.com.ua", extract_rozetka_html),
        ("shopgoodwill.com", extract_shopgoodwill_html),
    )
    for pattern, extractor in complex_extractors:
        if pattern in hostname:
            return extractor(html_text, page_url)
    if "bandcamp.com" in hostname:
        return _extract_bandcamp(html_text, page_url)

    profile = get_store_profile(page_url)
    if profile is None:
        return None
    return _extract_profile(profile, html_text, page_url)


def _extract_profile(
    profile: StoreProfile,
    html_text: str,
    page_url: str,
) -> ExtractionResult:
    document = _parse_document(html_text)
    title = _select(document, profile.title)
    if title:
        for pattern in profile.title_cleanup:
            title = re.sub(pattern, "", title, flags=re.I).strip()
    if title and any(blocked in title.lower() for blocked in profile.blocked_titles):
        return ExtractionResult(warnings=[f"{profile.name}_blocked_page"])

    current_price = normalize_price(_select(document, profile.current_price))
    if not current_price:
        current_price = _first_regex_price(html_text, profile.current_price_regex)

    old_price = normalize_price(_select(document, profile.old_price)) if profile.old_price else None
    if not old_price:
        old_price = _first_regex_price(html_text, profile.old_price_regex)

    if old_price and current_price:
        price, discount_price, has_discount = normalize_discount(old_price, current_price)
    else:
        price, discount_price, has_discount = current_price, None, False

    currency = normalize_currency(_select(document, profile.currency)) if profile.currency else None
    if not currency and profile.currency_regex:
        match = re.search(profile.currency_regex, html_text, re.I)
        currency = normalize_currency(match.group(1)) if match else None
    if not currency and price:
        currency = profile.default_currency

    image = normalize_image_url(_select(document, profile.image), page_url) if profile.image else None
    if image:
        for pattern, replacement in profile.image_replacements:
            image = re.sub(pattern, replacement, image, flags=re.I)

    product = ProductData(
        title=clean_text(title),
        description=clean_text(_select(document, profile.description))
        if profile.description
        else None,
        image=image,
        price=price,
        discount_price=discount_price,
        has_discount=has_discount,
        currency=currency,
    )
    source_name = f"store:{profile.name}"
    sources = {
        field_name: source_name
        for field_name, value in product.model_dump().items()
        if value not in (None, False)
    }
    if has_discount:
        sources["has_discount"] = source_name
    return ExtractionResult(product=product, sources=sources)


def _extract_bandcamp(html_text: str, page_url: str) -> ExtractionResult:
    document = _parse_document(html_text)
    title = None
    description = None
    image = None
    price = None
    currency = None

    for script in document.xpath("//script[@type='application/ld+json']/text()"):
        try:
            payload = json.loads(script)
        except (json.JSONDecodeError, TypeError):
            continue
        if not isinstance(payload, dict) or payload.get("@type") not in {
            "MusicAlbum",
            "MusicRecording",
        }:
            continue
        title = clean_text(payload.get("name"))
        description = clean_text(payload.get("description"))
        if description:
            description = re.sub(r"^PREVIEWS:\s*https?://\S+\s*", "", description, flags=re.I)
        raw_image = payload.get("image")
        if isinstance(raw_image, str):
            image = normalize_image_url(
                re.sub(r"_\d+(\.\w+)$", r"_10\1", raw_image),
                page_url,
            )
        for release in payload.get("albumRelease") or []:
            if not isinstance(release, dict):
                continue
            offers = release.get("offers")
            for offer in offers if isinstance(offers, list) else [offers]:
                if isinstance(offer, dict) and offer.get("price") is not None:
                    price = normalize_price(offer.get("price"))
                    currency = normalize_currency(offer.get("priceCurrency"))
                    break
            if price:
                break
        break

    if not price:
        match = re.search(r'"minimum_price"\s*:\s*([\d.]+)', html_text)
        price = normalize_price(match.group(1)) if match else None

    product = ProductData(
        title=title,
        description=description,
        image=image,
        price=price,
        currency=currency,
    )
    sources = {
        field_name: "store:bandcamp"
        for field_name, value in product.model_dump().items()
        if value not in (None, False)
    }
    return ExtractionResult(product=product, sources=sources)


def _parse_document(html_text: str) -> html.HtmlElement:
    parser = html.HTMLParser(encoding="utf-8", recover=True, huge_tree=False)
    try:
        return html.fromstring(html_text, parser=parser)
    except (etree.ParserError, ValueError):
        return html.fromstring("<html></html>", parser=parser)


def _select(document: html.HtmlElement, rule: FieldRule | None) -> str | None:
    if rule is None:
        return None
    for xpath in rule.xpaths:
        nodes = document.xpath(xpath)
        if not nodes:
            continue
        node = nodes[0]
        if isinstance(node, str):
            value = clean_text(node)
            if value:
                return value
            continue
        for attribute in rule.attributes:
            value = clean_text(node.get(attribute))
            if value:
                return value
        value = clean_text(node.text_content() if hasattr(node, "text_content") else node.text)
        if value:
            return value
    return None


def _first_regex_price(html_text: str, patterns: tuple[str, ...]) -> str | None:
    for pattern in patterns:
        match = re.search(pattern, html_text, re.I)
        if match:
            price = normalize_price(match.group(1))
            if price:
                return price
    return None

File: packages/scraper-core/src/stores/registry.ts
/**
 * Port of `services/scraper/app/extractors/stores/registry.py`, plus the two
 * HTML-only extractors that live in `enrichment.py` (`extract_rozetka_html`,
 * `extract_shopgoodwill_html`).
 */
import { selectValue, xpathFirst } from "../dom";
import { isRecord } from "../json";
import {
  cleanText,
  hostnameOf,
  normalizeCurrency,
  normalizeDiscount,
  normalizeImageUrl,
  normalizePrice,
} from "../normalization";
import { extractionOf, productFrom } from "../result";
import { type ExtractionResult, emptyExtraction } from "../types";
import {
  type ComplexExtractor,
  attributeFrom,
  extractAliexpress,
  extractAmazon,
  extractAsos,
  extractFlipkart,
  extractFoxtrot,
  extractN11,
  extractTarget,
  extractTrendyol,
  extractZalando,
} from "./complex";
import { extractEmbeddedStore, supportsEmbeddedStore } from "./embedded";
import { DEFAULT_FIELD_ATTRIBUTES, type FieldRule, PROFILES, type StoreProfile } from "./profiles";

const COMPLEX_EXTRACTORS: readonly (readonly [string, ComplexExtractor])[] = [
  ["amazon.", extractAmazon],
  ["aliexpress.", extractAliexpress],
  ["asos.com", extractAsos],
  ["foxtrot.com.ua", extractFoxtrot],
  ["n11.com", extractN11],
  ["target.com", extractTarget],
  ["flipkart.com", extractFlipkart],
  ["trendyol.com", extractTrendyol],
  ["zalando.", extractZalando],
  ["rozetka.com.ua", extractRozetkaHtml],
  ["shopgoodwill.com", extractShopgoodwillHtml],
];

function getStoreProfile(pageUrl: string): StoreProfile | null {
  const hostname = hostnameOf(pageUrl);
  return (
    PROFILES.find((profile) => profile.patterns.some((pattern) => hostname.includes(pattern))) ??
    null
  );
}

export function extractStoreProduct(
  document: Document,
  htmlText: string,
  pageUrl: string,
): ExtractionResult | null {
  const hostname = hostnameOf(pageUrl);

  if (supportsEmbeddedStore(pageUrl)) {
    const embedded = extractEmbeddedStore(document, pageUrl);
    if (embedded.product.title || embedded.product.price || embedded.product.image) {
      return embedded;
    }
  }

  for (const [pattern, extractor] of COMPLEX_EXTRACTORS) {
    if (hostname.includes(pattern)) return extractor(document, htmlText, pageUrl);
  }

  if (hostname.includes("bandcamp.com")) return extractBandcamp(document, htmlText, pageUrl);

  const profile = getStoreProfile(pageUrl);
  if (profile === null) return null;
  return extractProfile(profile, document, htmlText, pageUrl);
}

function select(document: Document, rule: FieldRule | undefined): string | null {
  if (!rule) return null;
  return selectValue(document, rule.xpaths, rule.attributes ?? DEFAULT_FIELD_ATTRIBUTES);
}

function extractProfile(
  profile: StoreProfile,
  document: Document,
  htmlText: string,
  pageUrl: string,
): ExtractionResult {
  let title = select(document, profile.title);
  if (title) {
    for (const pattern of profile.titleCleanup ?? []) {
      title = title.replace(resetRegex(pattern), "").trim();
    }
  }
  if (
    title &&
    (profile.blockedTitles ?? []).some((blocked) => title!.toLowerCase().includes(blocked))
  ) {
    return emptyExtraction([`${profile.name}_blocked_page`]);
  }

  let currentPrice = normalizePrice(select(document, profile.currentPrice));
  if (!currentPrice) currentPrice = firstRegexPrice(htmlText, profile.currentPriceRegex);

  let oldPrice = profile.oldPrice ? normalizePrice(select(document, profile.oldPrice)) : null;
  if (!oldPrice) oldPrice = firstRegexPrice(htmlText, profile.oldPriceRegex);

  const discount =
    oldPrice && currentPrice
      ? normalizeDiscount(oldPrice, currentPrice)
      : { price: currentPrice, discountPrice: null, hasDiscount: false };

  let currency = profile.currency ? normalizeCurrency(select(document, profile.currency)) : null;
  if (!currency && profile.currencyRegex) {
    const match = resetRegex(profile.currencyRegex).exec(htmlText);
    currency = match ? normalizeCurrency(match[1]) : null;
  }
  if (!currency && discount.price) currency = profile.defaultCurrency ?? null;

  let image = profile.image ? normalizeImageUrl(select(document, profile.image), pageUrl) : null;
  if (image) {
    for (const [pattern, replacement] of profile.imageReplacements ?? []) {
      image = image.replace(resetRegex(pattern), replacement);
    }
  }

  const product = productFrom({
    title: cleanText(title),
    description: profile.description ? cleanText(select(document, profile.description)) : null,
    image,
    price: discount.price,
    discount_price: discount.discountPrice,
    has_discount: discount.hasDiscount,
    currency,
  });
  return extractionOf(product, `store:${profile.name}`);
}

// --------------------------------------------------------------------------
// Bandcamp
// --------------------------------------------------------------------------

function extractBandcamp(document: Document, htmlText: string, pageUrl: string): ExtractionResult {
  let title: string | null = null;
  let description: string | null = null;
  let image: string | null = null;
  let price: string | null = null;
  let currency: string | null = null;

  const scripts = document.querySelectorAll("script[type='application/ld+json']");
  for (const script of Array.from(scripts)) {
    let payload: unknown;
    try {
      payload = JSON.parse(script.textContent ?? "");
    } catch {
      continue;
    }
    if (!isRecord(payload)) continue;
    const type = payload["@type"];
    if (type !== "MusicAlbum" && type !== "MusicRecording") continue;

    title = cleanText(payload.name);
    description = cleanText(payload.description);
    if (description) {
      description = description.replace(/^PREVIEWS:\s*https?:\/\/\S+\s*/i, "");
    }
    const rawImage = payload.image;
    if (typeof rawImage === "string") {
      image = normalizeImageUrl(rawImage.replace(/_\d+(\.\w+)$/, "_10$1"), pageUrl);
    }

    const releases = Array.isArray(payload.albumRelease) ? payload.albumRelease : [];
    for (const release of releases) {
      if (!isRecord(release)) continue;
      const offers = release.offers;
      const offerList = Array.isArray(offers) ? offers : [offers];
      for (const offer of offerList) {
        if (isRecord(offer) && offer.price !== undefined && offer.price !== null) {
          price = normalizePrice(offer.price);
          currency = normalizeCurrency(offer.priceCurrency);
          break;
        }
      }
      if (price) break;
    }
    break;
  }

  if (!price) {
    const match = /"minimum_price"\s*:\s*([\d.]+)/.exec(htmlText);
    price = match ? normalizePrice(match[1]) : null;
  }

  const product = productFrom({ title, description, image, price, currency });
  return extractionOf(product, "store:bandcamp");
}

// --------------------------------------------------------------------------
// Rozetka / ShopGoodwill (HTML halves of `enrichment.py`)
// --------------------------------------------------------------------------

function extractRozetkaHtml(
  document: Document,
  _htmlText: string,
  pageUrl: string,
): ExtractionResult {
  const current = enrichmentPrice(document, [
    "//*[contains(@class,'product-price__big')][1]",
    "//*[@data-testid='price'][1]",
    "//*[@itemprop='price'][1]",
  ]);
  const old = enrichmentPrice(document, [
    "//*[contains(@class,'product-price__small')][1]",
    "//*[contains(@class,'old-price')][1]",
    "//*[contains(@class,'price--old')][1]",
  ]);
  const discount =
    old && current
      ? normalizeDiscount(old, current)
      : { price: current, discountPrice: null, hasDiscount: false };

  const product = productFrom({
    title: enrichmentText(document, ["//h1[contains(@class,'product__title')][1]", "//h1[1]"]),
    description: attributeFrom(
      document,
      ["//meta[@property='og:description']", "//meta[@name='description']"],
      ["content"],
    ),
    image: normalizeImageUrl(
      attributeFrom(document, ["//meta[@property='og:image']"], ["content"]),
      pageUrl,
    ),
    price: discount.price,
    discount_price: discount.discountPrice,
    has_discount: discount.hasDiscount,
    currency: discount.price ? "UAH" : null,
  });
  return extractionOf(product, "store:rozetka");
}

function extractShopgoodwillHtml(
  document: Document,
  htmlText: string,
  pageUrl: string,
): ExtractionResult {
  let title =
    attributeFrom(document, ["//meta[@property='og:title']"], ["content"]) ??
    enrichmentText(document, ["//title", "//h1[1]"]);
  if (title) title = title.replace(/^(?:Used|New|Like New|Pre-Owned)\s+/i, "");

  let price: string | null = null;
  for (const pattern of [
    /"currentPrice"\s*:\s*([\d.]+)/i,
    /"minimumBid"\s*:\s*([\d.]+)/i,
    /"startingBid"\s*:\s*([\d.]+)/i,
    /"buyNowPrice"\s*:\s*([\d.]+)/i,
  ]) {
    const match = pattern.exec(htmlText);
    if (!match) continue;
    price = normalizePrice(match[1]);
    if (price) break;
  }

  const product = productFrom({
    title: cleanText(title),
    description: attributeFrom(
      document,
      ["//meta[@property='og:description']", "//meta[@name='description']"],
      ["content"],
    ),
    image: normalizeImageUrl(
      (attributeFrom(document, ["//meta[@property='og:image']"], ["content"]) ?? "")
        .split("\\")
        .join("/"),
      pageUrl,
    ),
    price,
    currency: price ? "USD" : null,
  });
  return extractionOf(product, "store:shopgoodwill_html");
}

/** Port of `enrichment.py::_text` — text content only, no `content` attribute. */
function enrichmentText(document: Document, expressions: readonly string[]): string | null {
  for (const expression of expressions) {
    const node = xpathFirst(document, expression);
    if (node === null) continue;
    const value = cleanText(typeof node === "string" ? node : node.textContent);
    if (value) return value;
  }
  return null;
}

/** Port of `enrichment.py::_price` — `content`, then `value`, then text. */
function enrichmentPrice(document: Document, expressions: readonly string[]): string | null {
  for (const expression of expressions) {
    const node = xpathFirst(document, expression);
    if (node === null) continue;
    const raw =
      typeof node === "string"
        ? node
        : node.getAttribute("content") || node.getAttribute("value") || node.textContent;
    const price = normalizePrice(raw);
    if (price) return price;
  }
  return null;
}

// --------------------------------------------------------------------------

function firstRegexPrice(htmlText: string, patterns: readonly RegExp[] | undefined): string | null {
  for (const pattern of patterns ?? []) {
    const match = resetRegex(pattern).exec(htmlText);
    if (!match) continue;
    const price = normalizePrice(match[1]);
    if (price) return price;
  }
  return null;
}

/** Fresh instance so a shared `g`-flagged literal never carries `lastIndex`. */
function resetRegex(pattern: RegExp): RegExp {
  return new RegExp(pattern.source, pattern.flags);
}
